arch/risc-v/src/mpfs/mpfs_can.c

```diff
@@ -0,0 +1,2953 @@
+
+  if (expect_true(!priv->filter.use_mask_filter
+      && !priv->filter.use_range_filter)
+      || (priv->filter.use_mask_filter
+      && (can_id_masked & priv->filter.mask) == priv->filter.code)
+      || (priv->filter.use_range_filter
+      && can_id_masked >= priv->filter.range_lower
+      && can_id_masked <= priv->filter.range_upper))
+  {
+    goto accepted;
+  }
+
+  return CAN_ERR;
+
+accepted:
+#endif  /* CONFIG_NETDEV_CAN_FILTER_IOCTL */
+
+  /* DLC */
+
+  dlc = (pmsg->msg_ctrl & MPFS_CAN_RX_MSG_CTRL_CMD_DLC)
+         >> MPFS_CAN_RX_MSG_CTRL_CMD_DLC_SHIFT;
+  if (dlc <= 8)
+    {
+     cf->can_dlc = dlc;
+    }
+  else
+    {
+      canerr("DLC = %d is out of range\n", dlc);
+      return CAN_ERR;
+    }
+
+  /* Data (big endian) */
+
+  for (i = 0; i < dlc; i++)
+    {
+      if (i < 4)
+        {
+          cf->data[i] = (uint8_t)((pmsg->data_high >> ((3 - i) * 8)) & 0xff);
+        }
+      else
+        {
+          cf->data[i] = (uint8_t)((pmsg->data_low >> ((7 - i) * 8)) & 0xff);
+        }
+    }
+
+  return CAN_OK;
+}
+
+/****************************************************************************
+ * Name: mpfs_receive_work
+ *
+ * Description:
+ *  An interrupt was received indicating the availability of a new RX packet
+ *
+ * Input Parameters:
+ *  priv  - Pointer to the private CAN driver state structure
+ *
+ * Returned Value:
+ *  None
+ *
+ * Assumptions:
+ *  Global interrupts are disabled by interrupt handling logic.
+ *
+ ****************************************************************************/
+
+static void mpfs_receive_work(void *arg)
+{
+  mpfs_can_instance_t *priv = (mpfs_can_instance_t *)arg;
+
+#ifdef CONFIG_DEBUG_CAN_INFO
+  caninfo("CAN RX interrupt received\n");
+#endif
+
+  while (CAN_VALID_MSG == mpfs_can_get_message(priv))
+    {
+      struct can_frame *cf = (struct can_frame *)priv->rxdesc;
+
+      /* Retrieve the CAN 2.0B frame */
+
+      if (CAN_OK != mpfs_can_retrieve_rx_frame(priv, cf))
+        {
+          /* Didn't receive full frame or message got filtered out */
+
+          continue;
+        }
+
+      /* Lock the network; we have to protect the dev.d_len, dev.d_buf
+      * and dev.d_iob from the devif_poll path
+      */
+
+      net_lock();
+
+      /* Copy the buffer pointer to priv->dev.d_buf  Set amount of data
+        * in priv->dev.d_len
+        */
+
+      priv->dev.d_len = sizeof(struct can_frame);
+      priv->dev.d_buf = (uint8_t *)cf;
+
+      /* Send to socket interface */
+
+      NETDEV_RXPACKETS(&priv->dev);
+      can_input(&priv->dev);
+
+      net_unlock();
+
+      /* Point the packet buffer back to the next Tx buffer that will be
+        * used during the next write.  If the write queue is full, then
+        * this will point at an active buffer, which must not be written
+        * to.  This is CAN_OK because devif_poll won't be called unless the
+        * queue is not full.
+        */
+
+      priv->dev.d_buf = (uint8_t *)priv->txdesc;
```

Comment:
Does this actually make sense? looks like some useless copy-paste from somewhere

Reply:
yes, this is required for switching the network device pointer back to the txdesc buffer just to make extra sure that it's not in rxdesc.. . If you check other nuttx CAN driver, this is a common practice